tlvc/arch/i86/drivers/block/directhd.c

```diff
@@ -680,14 +665,25 @@ int directhd_open(struct inode *inode, struct file *filp)
     if (((int) hd[minor].start_sect) == -1)	/* FIXME is this initialized */
 	return -ENXIO;
 
-    if (!S_ISCHR(inode->i_mode)) 	/* Don't count raw opens */
+#if defined(USE_LOCALBUF) || defined(CONFIG_FS_XMS_BUFFER)
+    if (!(localbuf = heap_alloc(BLOCK_SIZE, HEAP_TAG_DRVR))) {
+	printk("athd: cannot allocate bounce buffer\n");
+            return -EBUSY;
+    }
```

Comment:
Doesn't this alloc need to be happen only once when open_drv_count == 0? This seems to allocate a new bounce buffer into a single global each time the driver is opened - which occurs on separate raw opens and possibly other opens, correct?

No mem should return -ENOMEM, is -EBUSY for another reason?

A heap allocated bounce buffer has a (small but real) chance of major DMA overwrite bug if/when the kernel DS segment spans a 64k boundary, so this method is unsafe on early machines.

Reply:
Thanks, rushing things is never good. Appreciated. 
There is no DMA at work here, but the xd driver has borrowed some code from the floppy driver and needs some work in the dma dept.

Reply:
Oh, I had thought you mentioned that your HD driver used DMA in a previous PR deleting some DMA stuff.

Reply:
I did, although not this driver but the older (xt generation) mfm (xd) driver.
 Most ide drives and controllers can do dma but thanks to on-drive buffers, pio is faster on old hw.